src/vs/workbench/electron-sandbox/window.ts

```diff
@@ -806,6 +806,74 @@ export class NativeWindow extends BaseWindow {
 		});
 	}
 
+	private async openTunnel(address: string, port: number): Promise<RemoteTunnel | string | undefined> {
+		const remoteAuthority = this.environmentService.remoteAuthority;
+		const addressProvider: IAddressProvider | undefined = remoteAuthority ? {
+			getAddress: async (): Promise<IAddress> => {
+				return (await this.remoteAuthorityResolverService.resolveAuthority(remoteAuthority)).authority;
+			}
+		} : undefined;
+		const tunnel = await this.tunnelService.getExistingTunnel(address, port);
+		if (!tunnel || (typeof tunnel === 'string')) {
+			return this.tunnelService.openTunnel(addressProvider, address, port);
+		}
+		return tunnel;
+	}
+
+	async resolveExternalUri(uri: URI, options?: OpenOptions): Promise<IResolvedExternalUri | undefined> {
+		let queryTunnel: RemoteTunnel | string | undefined;
+		let resolved = undefined;
+		if (options?.allowTunneling) {
+			const portMappingRequest = extractLocalHostUriMetaDataForPortMapping(uri);
+			const queryPortMapping = extractQueryLocalHostUriMetaDataForPortMapping(uri);
+			if (queryPortMapping) {
+				queryTunnel = await this.openTunnel(queryPortMapping.address, queryPortMapping.port);
+				if (queryTunnel && (typeof queryTunnel !== 'string')) {
+					resolved = uri.with({ query: uri.query.replace(`${queryPortMapping.address}:${queryPortMapping.port}`, queryTunnel.localAddress) });
```

Comment:
Why does the query need to be replaced?

Reply:
I think I see why it needs to be replaced: If the forward ends up available locally at a different address then the query needs to be updated to point to that new local address.

Reply:
Right.